hs-bindgen/src/HsBindgen/Backend/Common.hs

```diff
@@ -30,10 +33,16 @@ class BackendRep be where
   type Decl be :: Type
   type Ty   be :: Type
 
-  resolve :: be -> Global   -> Name be  -- ^ Resolve name
-  mkExpr  :: be -> SExpr be -> Expr be  -- ^ Construct expression
-  mkDecl  :: be -> SDecl be -> Decl be  -- ^ Construct declaration
-  mkType  :: be -> SType be -> Ty   be  -- ^ Construct type
+  -- | Resolve a global name.
+  resolve      :: be -> Global -> Name be
+  -- | Resolve a local name.
+  resolveLocal :: be -> HsName NsVar -> Name be
```

Comment:
OK but that was my question: what do I use for e.g. the name of a user-defined macro, e.g. the occurence of `FOO` within the RHS of `BAR` in:

```c
#define FOO(X,Y) X + Y
#define BAR(X,Y) 2 * FOO(X,Y)
```

Do I add another field to `Global`:

```hs
data Global
  = ...
  | UserGlobal (HsName NsVar)
```

or something?

Reply:
AFAICT, (@edsko should verify, as it's his creation, I'm only assuming) we use constructors elsewhere:

```
data SExpr be =
    EGlobal Global
  | EVar (Fresh be Bound)  -- ^ local
  | ECon (HsName NsConstr) -- ^ kind of global, but refers to "our" names.
```

I.e. `Global` are valid in all contexts / modules. More like built-ins

Reply:
And to add, it's not sure why we have `Backend` class.

In `Backend.Common.Translation`, we genetate `SDecl`, `SExpr` etc. See e.g

```
instance Backend be => ToBE be Hs.Decl where
  type Rep be Hs.Decl = Decl be
  toBE be (Hs.DeclData d) = mkDecl be <$> toBE be d
  toBE be (Hs.DeclNewtype n) = mkDecl be <$> return (newtypeToBE be n)
  toBE be (Hs.DeclInstance i) = inst be <$> toBE be i
```

`inst` is also essentially a `mkDecl`. Why we don't simply have a `Hs.Decl -> SDecl` function escapes me.

The only truly backend specific thing we do at that step is generate names, but I'd argue we could delay that too, making `SDecl`, `SExpr` backend independent (Haskell AST).

Reply:
Yes, I think I agree with Oleg's understanding in https://github.com/well-typed/hs-bindgen/pull/270#discussion_r1837007175 .

As for the simplification proposed in https://github.com/well-typed/hs-bindgen/pull/270#discussion_r1837874453 , I'm not opposed to that. I went through a few iterations of this setup, perhaps I ended up on this one more due to historical accident than anything else. If it can be made simpler, let's make it simpler.